java-checks/src/main/java/org/sonar/java/checks/RedundantNullabilityAnnotationsCheck.java

```diff
@@ -0,0 +1,120 @@
+/*
+ * SonarQube Java
+ * Copyright (C) 2012-2024 SonarSource SA
+ * mailto:info AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.java.checks;
+
+import java.util.Arrays;
+import java.util.List;
+import java.util.Objects;
+import java.util.Optional;
+import org.sonar.check.Rule;
+import org.sonar.plugins.java.api.IssuableSubscriptionVisitor;
+import org.sonar.plugins.java.api.semantic.SymbolMetadata;
+import org.sonar.plugins.java.api.tree.ClassTree;
+import org.sonar.plugins.java.api.tree.MethodTree;
+import org.sonar.plugins.java.api.tree.Tree;
+import org.sonar.plugins.java.api.tree.VariableTree;
+
+import static org.sonar.java.checks.helpers.NullabilityDataUtils.nullabilityAsString;
+import static org.sonar.plugins.java.api.semantic.SymbolMetadata.NullabilityLevel.CLASS;
+import static org.sonar.plugins.java.api.semantic.SymbolMetadata.NullabilityLevel.METHOD;
+import static org.sonar.plugins.java.api.semantic.SymbolMetadata.NullabilityLevel.PACKAGE;
+import static org.sonar.plugins.java.api.semantic.SymbolMetadata.NullabilityLevel.VARIABLE;
+
+@Rule(key = "S6665")
+public class RedundantNullabilityAnnotationsCheck extends IssuableSubscriptionVisitor {
+
+  private static final String ISSUE_MESSAGE = "Remove redundant annotation %s as inside scope annotation %s.";
+
+  @Override
+  public List<Tree.Kind> nodesToVisit() {
+    return Arrays.asList(Tree.Kind.INTERFACE, Tree.Kind.CLASS, Tree.Kind.RECORD);
+  }
+
+  @Override
+  public void visitNode(Tree tree) {
+    ClassTree classTree = (ClassTree) tree;
+    // check if outer class - this will return default package if necessary
+    if (Objects.requireNonNull(classTree.symbol().owner()).isPackageSymbol()) {
```

Comment:
This line cuts the visitation of any inner classes, but they may have their own NullMarked / NullUnmarked annotations. The following test cases fail due to this:

```java
@NullUnmarked
class InnerNullmarked {
  @NullMarked
  class Inner {
    @NonNull Object o; // Noncompliant
  }
}

@NullMarked
class InnerNullUnmarked {
  @NullUnmarked
  class Inner {
    @NonNull Object o; // Compliant
  }
}

@NullUnmarked
class InnerRedundantNullMarked {
  @NullMarked
  class Inner {
    @NullMarked // Noncompliant
    class InnerInner {}
  }
}

@NullMarked
class InnerRedundantNullUnmarked {
  @NullUnmarked
  class Inner {
    @NullUnmarked // Noncompliant
    class InnerInner {}
  }
}
```